core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala

```diff
@@ -191,7 +191,20 @@ private[spark] class PythonWorkerFactory(pythonExec: String, envVars: Map[String
         daemon = pb.start()
 
         val in = new DataInputStream(daemon.getInputStream)
-        daemonPort = in.readInt()
+        try {
+          daemonPort = in.readInt()
+        } catch {
+          case exc: EOFException =>
+            throw new IOException(s"No port number in $daemonModule's stdout")
+        }
+
+        // test that the returned port number is within a valid range.
+        // note: this does not cover the case where the port number
+        // is arbitrary data but is also coincidentally within range
+        if (daemonPort < 1 || daemonPort > 0xffff) {
```

Comment:
Why don't we just copy the condition in `java.net.InetSocketAddress.checkPort`: `if (port < 0 || port > 0xFFFF)`?

Reply:
Port 0 has special meaning. A program passes port 0 when it wants the system to choose an unused port on the program's behalf. So, the daemon should not return 0.

It's valid to pass port 0 to InetSocketAddress, since you might be asking for the system to assign a port for you.

However, following my own logic, the code in my pull request really should be checking for the range 49152-65535 (ephemeral range) instead of 1-65535, but I didn't have the nerve to make it that restrictive.

Reply:
I am saying we can just safely move the same check and fail fast, which is simple and theoretically safe.

Reply:
Sorry @HyukjinKwon , I didn't quite get the last point. Could you rephrase?

Reply:
I mean, I left my sign-off because what we do is basically move the _same_ check (`java.net.InetSocketAddress.checkPort`) ahead and another one is simply to wraps an exception, `EOFException`. I think we are here safe in theory.

I got your point of reserved ports and now the condition became narrower. I should check other things like which error it produces before in this case and if the current error message is nicer. Also, this seems not completely addressing the concerns about it.

I was wondering if this is worth doing these stuff. If you strongly prefer this, I won't stay against but may request few more investigations.

Reply:
Oh, I see. Let me address the two parts of the comment separately:

First part: What's the point of throwing an exception for a bad port number when the original handling did that already?

The original handling was:

<pre>
 java.lang.IllegalArgumentException: port out of range:1315905645
        at java.net.InetSocketAddress.checkPort(InetSocketAddress.java:143)
</pre>

This error occurred in a different function than the one that obtained the port number. So you had to track down the source of the port number. This actually added an extra step to the original debugging for the sitecustomize.py issue.

The proposed handling is:

<pre>
java.io.IOException: Bad port number in pyspark.daemon's stdout: 0x4e6f206d
        at org.apache.spark.api.python.PythonWorkerFactory.startDaemon(PythonWorkerFactory.scala:205)
</pre>

Here we're not saying "somehow we got a bad port number", but that a particular python module (the name of which is displayed, since the name of that module is configurable) has returned bad data.

Also, the check is at the point at which the port number is obtained, not waiting until sometime later in another function where the port number is used (and where possibly something else has changed the port number, which is not likely, but you would need to check that).

Perhaps the message could be better, e.g.,:

<pre>
Bad data in pyspark.daemon's output.
Expected valid port number, got 0x4e6f206d.
PYTHONPATH set to '/Users/brobbins/github/spark_fork/python/lib/pyspark.zip:/Users/brobbins/github/spark_fork/python/lib/py4j-0.10.6-src.zip:/Users/brobbins/github/spark_fork/assembly/target/scala-2.11/jars/spark-core_2.11-2.4.0-SNAPSHOT.jar:/Users/brobbins/github/spark_fork/python/lib/py4j-0.10.6-src.zip:/Users/brobbins/github/spark_fork/python/:'
Command to run python module was 'python -m pyspark.daemon'
Check whether you have a sitecustomize.py module that may be printing output to stdout.
</pre>

Second part: Why don't we fix this?

That's reasonable. A couple of points:

- The name of the python daemon module is now configurable so that the module can be wrapped with customizations. It appears that this is only in the main branch and not even released on Spark 2.3, so it might be safe to change daemon.py (and potentially its existing wrappers) to return the port number in a different way.
- I don't have a good feel for how often sitecustomize.py is used, so not sure of the relative value of some mild hacking up of this code vs. just letting the user know what happened.

Reply:
yeah I could go either way on this.  Personally i think just the enhanced error message you have above would be useful, without going through the added trouble of using another approach.  I'll defer to your opinion @HyukjinKwon

Reply:
Thanks, @squito. My only worry is, what is the error message like befer / after this change when `daemonPort` happens to be 0 because it was filled in pyspark.daemon's stdout unexpectedly by the sitecustomize or usercustomize whatever?

I assume the error message becomes different as you described above because `java.net.InetSocketAddress.checkPort` passes `0` case but your proposed change also catches `0` case too for the reason you described above.

This case `0` is different because the previous codes probably try to attempt a connection whether it's a valid port or not, after passing `java.net.InetSocketAddress.checkPort`, if I didn't misunderstand.

I want to be very sure on this. Is it guranteed to be always failed without any hole and the error message is always better in this case? How about other ports reserved or "well-known" ? Should we really take this into account here?

Please let me know if I missed something.

Reply:
ah I see, I think you are worried about something other than what bruce and I thought.  Your concern is that we might throw an exception for some values that are actually perfectly legitimate.  Port 0 being special is a pretty standard thing -- its mentioned in the constructor for ServerSocket: https://docs.oracle.com/javase/7/docs/api/java/net/ServerSocket.html#ServerSocket%28int%29

which implies that you shouldn't ever open a Socket on port 0, though I don't see that officially documented.  At least on my laptop, I get different errors if I try to connect to port 0, vs. just connecting to a bogus port:

```scala
scala> val s2 = new Socket("localhost", 1234)
java.net.ConnectException: Connection refused
  at java.net.PlainSocketImpl.socketConnect(Native Method)
  at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
  at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
  at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
  at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
  at java.net.Socket.connect(Socket.java:589)
  at java.net.Socket.connect(Socket.java:538)
  at java.net.Socket.<init>(Socket.java:434)
  at java.net.Socket.<init>(Socket.java:211)
  ... 29 elided

scala> val s3 = new Socket("localhost", 0)
java.net.NoRouteToHostException: Can't assign requested address
  at java.net.PlainSocketImpl.socketConnect(Native Method)
  at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
  at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
  at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
  at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
  at java.net.Socket.connect(Socket.java:589)
  at java.net.Socket.connect(Socket.java:538)
  at java.net.Socket.<init>(Socket.java:434)
  at java.net.Socket.<init>(Socket.java:211)
  ... 29 elided
```

so I think its pretty safe to say that daemon.py (or whatever) shouldn't be passing back `0` as the port to bind to.

Still -- it is *clearly* safer to instead have the port written to some other file, or (another) socket, so that you we wouldn't have to worry about the details of this error handling.

Reply:
Ah, OK. Thanks for clarification. Maybe, I was caring too much about it. Thanks all for bearing with me. I am fine as is.

Reply:
> Still -- it is clearly safer to instead have the port written to some other file, or (another) socket, so that you we wouldn't have to worry about the details of this error handling.

Let's go that route. I will have a PR in the coming days.